[AWSX-537] Support python version 3.10

ge0Aja opened the pull request:
### What does this PR do?

Support python version 3.10

### Motivation

Support request

### Testing Guidelines

<!--- How did you test this pull request? --->

### Additional Notes

This PR adds support for python version 3.10  which requires a higher version for `datadog-lambda` package.
The usage of the higher version for that package adds a new field `DD-API-KEY` in the header section of the of the events response forwarded by the function. It also removes the API key string from the `path` parameter.
 

### Types of changes

- [ ] Bug fix
- [ ] New feature
- [x] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [x] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [x] This PR passes the integration tests (ask a Datadog member to run the tests)
- [x] This PR passes the unit tests 
- [x] This PR passes the installation tests (ask a Datadog member to run the tests)

ge0Aja committed:
Support python versions 3.10, 3.11

ge0Aja committed:
Remove a wrong rm statement from Dockerfile

ge0Aja committed:
Fix python versions

ge0Aja committed:
Update Itests

ge0Aja committed:
Use strings for python versions in in yaml config files.

Python 3.10 version is seen as 3.1 in github actions.
See https://github.com/actions/runner/issues/1989

ge0Aja commented:
had to use python versions as strings for triggering GH actions 
https://github.com/actions/runner/issues/1989

ge0Aja committed:
cleanup comments

ge0Aja committed:
Update readme

tyrcho commented on aws/logs_monitoring/tools/integration_tests/external_lambda/serverless.yml:
```diff
@@ -1,7 +1,7 @@
 provider:
   name: aws
-  runtime: python2.7
+  runtime: python3.11
```
is this file really used ? I'm surprised in was on 2.X

ge0Aja replied:
caught that when I was doing the version changes, not sure if it is used.

tyrcho commented on aws/logs_monitoring/README.md:
```diff
@@ -138,6 +138,10 @@ If you can't install the Forwarder using the provided CloudFormation template, y
 
+### Upgrade an older version to +3.97.0
+
+Since version 3.97.0 the Lambda function has been updated to require **Python 3.11**. If upgrading an older forwarder installation to 3.97.0 or above, ensure the AWS Lambda function is configured to use Python 3.11
```
it seems from most other files that we actually support 3.10 and 3.11 which is not really clear in this doc

ge0Aja replied:
we can update the readme to include both versions, I followed the convention in the readme file

tyrcho commented on aws/logs_monitoring/tools/Dockerfile_bundle:
```diff
@@ -19,4 +19,3 @@ RUN rm -rf ./botocore*
 # like `os.execl`, which cause security scans to fail for certain customers.
 # These files are not directly used by the Forwarder.
 RUN rm ./ddtrace/commands/ddtrace_run.py
-RUN rm ./decorator.py
```
why this change ?

ge0Aja replied:
file not found error was thrown during ITests, it seems it was added during a dependency version bump which created this file.
Now, it is not created, so rm is failing

tyrcho commented on aws/logs_monitoring/tools/add_new_region.sh:
```diff
@@ -12,7 +12,7 @@ set -e
 OLD_REGION='us-east-1'
 
-PYTHON_VERSIONS_FOR_AWS_CLI=("python3.9")
+PYTHON_VERSIONS_FOR_AWS_CLI=("python3.11")
```
do we support only one version here ? why an array then ?

ge0Aja replied:
it seems we support only the highest version, not sure why

ge0Aja committed:
Support Python 3.10

Support Python version 3.10 for first version bump before adding support for Python 3.11

ge0Aja committed:
Update python version range

RaphaelAllier commented on .github/workflows/lint.yml:
```diff
@@ -11,7 +11,7 @@ jobs:
         uses: actions/setup-python@v3
         with:
-          python-version: 3.9
+          python-version: '3.10'
```
Why do we add quotes here?

RaphaelAllier commented on .github/workflows/test.yml:
```diff
@@ -8,7 +8,7 @@ jobs:
       max-parallel: 4
       matrix:
-        python-version: [3.8, 3.9]
+        python-version: ['3.9', '3.10']
```
Same question, why do we use quotes now ?

ge0Aja replied:
see https://github.com/actions/runner/issues/1989, same for the above

RaphaelAllier commented on aws/logs_monitoring/README.md:
```diff
@@ -138,6 +138,10 @@ If you can't install the Forwarder using the provided CloudFormation template, y
 
 If you encounter issues upgrading to the latest version, check the Troubleshooting section.
 
+### Upgrade an older version to +3.97.0
```
You need to update the cloudformation template we use to deploy the forwarder to ensure we use Python 3.10 as well

RaphaelAllier commented on aws/logs_monitoring/setup.py:
```diff
@@ -12,12 +12,12 @@
-    python_requires=">=3.7, <3.10",
-    install_requires=["datadog-lambda==3.39.0", "requests-futures==1.0.0"],
+    python_requires=">=3.9, <3.11",
+    install_requires=["datadog-lambda==4.77.0", "requests-futures==1.0.0"],
```
Where's that version change coming from ?

ge0Aja replied:
required dependency version update to support versions 3.10 -> 3.11

RaphaelAllier commented on aws/logs_monitoring/tools/integration_tests/recorder/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9
+FROM python:3.11
```
This still mentions 3.11 (PR title states 3.10)

RaphaelAllier commented on aws/logs_monitoring/tools/publish_layers.sh:
```diff
@@ -13,7 +13,7 @@ set -e
 trap "pkill -P $$; exit 1;" INT
 
-PYTHON_VERSIONS_FOR_AWS_CLI=("python3.9")
+PYTHON_VERSIONS_FOR_AWS_CLI=("python3.11")
```
Still mentions 3.11 vs 3.10 in title of the PR

ge0Aja committed:
Fix CLI & Cloudformation python versions

ge0Aja committed:
Fix Itests

- Use the correct python runtime for tester and recorder docker images
- Add a health check for the recorder service in the docker-compose file
- Update service dependency
- Pass the correct python runtime version when bundling the forwarder code for the Itests
- Remove unused docker files as we are using the image version in the compose file
